391 Sequence without confronting views - Autre implémentation

Francois389 opened the pull request:
close #391 

Une séquence peut se jouer sans phase d'évaluation par les pairs.

Configuration de la séquence, sans la deuxième phase
![image](https://github.com/user-attachments/assets/4ccb205a-590a-4272-915b-0cfb7a35e9fc)

Step et Command sans la phase 2
![image](https://github.com/user-attachments/assets/1d995dbb-4aee-45c2-ab3f-fbcd74d588d6)

Le Dahsboard qui n'affiche qu'une seule phase
![image](https://github.com/user-attachments/assets/9371263d-1bd8-4098-b3fa-aad62ffd1918)

Pour ce faire, j'ai empêcher la création de l'interaction Evaluation.

Francois389 committed:
feat: initialise props in Thymleaf

Francois389 committed:
fix: clean

Francois389 committed:
feat: update component so we can deactivte Phase 2

Francois389 committed:
feat: update thymleaf template to give the necessary props to ui component

Francois389 committed:
feat(vue): update event data

Francois389 committed:
feat: extract value for phase 2 config to new class

Francois389 committed:
feat: use correct endpoint

Francois389 committed:
feat: use EvaluationPhaseConfig to pass value instead of literal

+ format, clean and document code

Francois389 committed:
feat: update second phase form to asynchronous request

+ fix duplicate reload script

Francois389 committed:
test: add test to CommandModelFactory

Francois389 committed:
feat: remove initialisation of evaluation phase if needed

WIP CommandModel

Francois389 committed:
feat: update Command, Step and Dashboard logic for when there is no second phase

Francois389 committed:
feat: clean code and fix test

Francois389 committed:
feat: enable CHAT_EVALUATION feature in test

Francois389 committed:
test: add test

test: add test to LearnerPhaseFactory

test: add test to LearnerPhaseService

test: add test to StepsModelFactory

test: add test to StepsModel

test: add test to Interaction

+ remove unused parameter 'user'
+ remove unused function 'hasAnyResult'

jtranier commented on server/src/main/kotlin/org/elaastic/player/command/CommandModel.kt:
```diff
@@ -25,13 +25,16 @@ data class CommandModel(
     val statementId: Long,
     val interactionId: Long?,
     val interactionRank: Int?,
+    val nextInteractionRank: Int?,
+    val nextNextInteractionRank: Int?,
```
Je n'ai pas compris l'intérêt de `next...` et `nextNext...` ? 
Les phases d'une séquence (avant on parlait d'interaction parce qu'il y avait confusion entre phase & interaction) doivent  être numérotées par leur position dans la séquence. `1, 2, 3` s'il y en a 3 et `1, 2` s'il y en a 2.

On ajoute le fait que chaque phase est d'un certain type.

Francois389 replied:
Il me serve à fournir le numéro de la phase à Thymleaf pour afficher le bon message :
```html
<span th:text="#{player.sequence.interaction.start(${commandModel.nextNextInteractionRank})}"></span>
```
![image](https://github.com/user-attachments/assets/9ae85e96-55c3-46d4-9b44-400a5359e23f)

jtranier replied:
Il me semble que next, c'est toujours n+1 si ça existe non ? 
Et me démarrer la phase n+2 (qui revient à sauter la phase suivante), sera toujours n+2. Il n'existe que s'il y a 2 phases à suivre, et peut être d'autres conditions éventuelles

Francois389 replied:
Un détail différent de la précédente implémentation, c'est que le numéro des phases se suivent.
Donc dans le cas d'une séquence à 2 phase en partant de la phase 1, 
je sais que la prochaine c'est `nextInteractionRank = 2`  donc j'affiche `Démarrer la phase 2`
et que la suivante c'est `nextNextInteractionRank = null` donc je n'affiche pas l'action `skip`


Ah, oui je viens de voir que mon image n'est pas correcte.

Francois389 replied:
Je vais changer le texte des actions pour ne plus utiliser les ranks des phases.

jtranier commented on server/src/main/kotlin/org/elaastic/player/dashboard/DashboardModelFactory.kt:
```diff
@@ -163,12 +161,10 @@ class DashboardModelFactory(
         learnerHasAnswered: Boolean,
         responsePhaseState: DashboardPhaseState
     ): LearnerStateOnPhase {
```
`LearnerStateOnPhase` sonne étrage. 
`LearnerPhaseState` me semble plus approprié

Francois389 replied:
Corrigé avec 0b71bf1beb6b17662f275d97f01a832993d1044c

jtranier commented on server/src/main/kotlin/org/elaastic/player/dashboard/DashboardPhaseState.kt:
```diff
@@ -0,0 +1,66 @@
+    COMPLETED,
+
+    /** The phase is not enabled */
+    NONE
```
Intuitivement le fait que l'on ait besoin d'un `NONE` ici me fait penser qu'il y a un défaut de conception. Le point étant complexe, je pense qu'il faut que l'on en discute de vive voix.

J'ai vu que dans le Dashboard, il y a une `phase1` / `phase2` en dur. C'était logique quand on imposait statiquement 3 phases toujours du même type, mais là, il faut probablement repenser le dashboard de manière plus générique. 
De la même manière que la `Sequence`, pour laquelle on a simplement enlevé la phase de confrontation de point de vue quand on ne la veut pas.

Francois389 replied:
Je suis d'accord, 
mais je pense que l'évolution du Dashboard vers quelque chose de plus générique, ne fais pas partie de cette évolution.

Francois389 replied:
Voir l'US #420

jtranier commented on server/src/main/kotlin/org/elaastic/player/sequence/SequenceController.kt:
```diff
@@ -19,14 +19,12 @@ class SequenceController(
         @PathVariable sequenceId: Long,
     ): String {
-        val user = (authentication.principal as PrincipalUserResolver).elaasticUser
         val sequence = sequenceService.get(sequenceId)
```
Ici il y a un problème de sécurité (qui était là avant ; tu le mets juste en évidence).
A partir du moment où l'on récupére une sequence avec `sequenceService.get`, il faut s'assurer que l'on est un utilisateur authentifié qui a le droit de faire ça

Francois389 replied:
Corrigé avec 0b71bf1beb6b17662f275d97f01a832993d1044c

jtranier commented on server/src/main/kotlin/org/elaastic/player/steps/StepsModel.kt:
```diff
@@ -4,17 +4,20 @@ import org.elaastic.player.dashboard.DashboardPhaseState
         ACTIVE,
-        COMPLETED;
+        COMPLETED,
+        NONE;
```
Pourquoi a-t-on besoin dun `PhaseState.NONE` ?

Francois389 replied:
Il me fallait un moyen de dire à Thymleaf de ne pas afficher une phase.
Mais comme tu l'as mentionné dans ton prochain commentaire, il faudrait une liste, ce qui enlèvera le besoin de cet état.

Francois389 replied:
Voir l'US #419

jtranier commented on server/src/main/kotlin/org/elaastic/player/steps/StepsModel.kt:
```diff
@@ -4,17 +4,20 @@ import org.elaastic.player.dashboard.DashboardPhaseState
 data class StepsModel(
-    val responseSubmissionState: PhaseState,
-    val evaluationState: PhaseState,
-    val readState: PhaseState,
+    val responseSubmission: PhaseStep?,
+    val evaluation: PhaseStep?,
+    val read: PhaseStep?,
```
Il faut reprendre le `StepsModel` maintenant... 
Avant c'était `response, evaluation, result`. 

Idéalement ça devrait être une liste de de `PhaseStep`.
Je ne sais pas dire sans me plonger vraiment dedans si on peut atteindre cet object tout de suite ou si on a besoin d'étape intermédiaire...

Francois389 replied:
C'est possible, mais pour avoir un truc totalement générique, il faut juste que je fournisse à Thymleaf des informations en plus :
- l'icone
- le nom de la phase `Phase 1`
- la description `Réponse argumenté`
- les statistiques s'il y en a

jtranier replied:
Complètement. Ce n'est pas trop compliqué à mettre en oeuvre pour les infos statiques.
Faut voir comment on gère les statistiques effectivemetn ... 
On peut même imaginer que le type de phase permette d'associer un template de steps spéficifique.

Ou bien... On se lance maintenant dans la réimplementation du Steps en composant Vue pour simplifier les choses. C'est un travail que @alexLcmt a commencé

Francois389 replied:
Ca https://67878cdbd8100262bdaab868-eemhbfjgkt.chromatic.com/?path=/docs/progressbar-progressbar--docs ?
C'est sur la branche [archive/feat/progressBar](https://github.com/elaastic/elaastic-questions-server/tree/archive/feat/progressBar)

Francois389 replied:
Je vais laisser le StepsModel comme ça et on le changera quand le composant Vue sera dev.

Francois389 replied:
Voir l'US #419

jtranier commented on server/src/main/kotlin/org/elaastic/sequence/phase/LearnerPhaseFactory.kt:
```diff
@@ -15,30 +16,19 @@ import org.springframework.stereotype.Service
+    ): LearnerPhase = when (interactionType) {
+        InteractionType.ResponseSubmission -> LearnerResponsePhase(learnerSequence, phaseIndex, active, state)
+        InteractionType.Evaluation ->
+            when (learnerSequence.sequence.evaluationMethod) {
```
`evaluationMethod` ne doit pas être portée par la `sequence` mais par la `evaluationPhaseConfig` qui est rattaché à l'une des phases de la config

Francois389 replied:
<details><summary>Removed</summary>
<p>
Avec 4cc64377588b70f4b02986dc7b8f65b5c2ac72a9 , j'ai débranché tous les appels à `evaluationMethod` pour utiliser `.getEvaluationSpecification().evaluationMethod`.

Mais il faut que je fasse la migration de la BD.
Et il faut que je réfléchisse comment je vais insérer la valeur dans un JSON.
</p>
</details> 

Cette évolution pourra faire l'objet d'une autre PR

Francois389 replied:
Voir l'US #418

jtranier commented on server/src/main/kotlin/org/elaastic/sequence/phase/LearnerPhaseFactory.kt:
```diff
@@ -15,30 +16,19 @@ import org.springframework.stereotype.Service
 class LearnerPhaseFactory {
 
     fun build(
-        phaseDescriptor: PhaseDescriptor,
+        interactionType: InteractionType,
```
Je ne pense pas qu'il faille remplacer `PhaseDescriptor` par `InteractionType`.

Au début d'Elaastic, il y a eu confusion entre "Phase" et "Interaction".
On est maintenant parti sur l'idée qu'il y a des phases qui s'enchainent (orchestrées par l'enseignant ou l'apprenant selon le mode), et, au sein d'une phase on pourra avoir une ou plusieurs interactions (ex : évaluer les réponses des pairs + changer sa propre réponse).

Pour construire une phase, on a donc besoin de savoir de quel type de phase il s'agit & d'avoir la configuration de cette phase.

Francois389 replied:
Pour avoir cette information du type de l'interaction, je me base sur les interactions de la séquence :
```kotlin
// LearnerPhaseService.loadPhaseList (l.38)
sequence.interactions.keys.forEachIndexed { index, interactionType ->
  // [...]
}
```
Alors que le PhaseDescriptor, était une liste de Phase qui était écrite en dur.

Ce changement me permet de construire les phase en fonction de la séquence.

jtranier replied:
Tu as raison sur le fait qu'il faut charger cette info depuis la base.
J'ai tiqué parce que l'on est entrain d'opérer un renommage de ces concepts et que, ici, c'est bien une phase que l'on construit et pas une interaction. Mais ce qu'une question de nommage / réorganisation des données et des types.

Je reviendrais sur ce point après.

Francois389 committed:
feat: add request change

- renamed `LearnerStateOnPhase` to `LearnerPhaseState`
- check if user can access config sequence modal

Francois389 committed:
feat: fix test

Francois389 committed:
feat(config-sequence): extract each phase config to a new component

feat(config-sequence): extract ResponsePhase config

feat(config-sequence): extract ConfrontingViewPhase config

feat(config-sequence): extract ResultPhase config

chore(config-sequence): refactor folder

Francois389 committed:
chore(ui-component): ui-component v1.0.9

Francois389 committed:
feat(config-sequence): update server with new SequenceConfig component

feat(config-sequence): update SequenceConfig to get the new form data

feat(config-sequence): insert start sequence form below the command and not in a modal

Commit:
todo: Add TODO to update start method signature

Francois389 committed:
feat(config-sequence): update SequenceConfig component

- hide nbEvaluationToEvaluate and evaluationMethod when studentGiveExplanation is false
- add notice for student change their response
- use `v-model`

Francois389 committed:
feat(i18n): remove unused i18n value

this value have been copied in a Vue component

Francois389 committed:
feat(config-sequence): remove old start method signature

jtranier committed:
Merge branch 'develop' into jtra/feat/391-sequence-without-confronting-views-2

# Conflicts:
#	ui-components/src/main.ts

jtranier committed:
refactor(BootstrapService): Improve readability

jtranier committed:
refactor(PlayerController): Improve readability

jtranier committed:
chore(SequenceService): Remove a deprecated TODO mark
